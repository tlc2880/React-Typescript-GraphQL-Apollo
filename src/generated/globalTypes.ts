//  Name: Tommy Cao
//  Date: 10/22/20
//  Title: Generating TypeScript Types from GraphQL Schema in Apollo

/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

/**
 * List of supported image content types.
 */
export enum ImageContentType {
  JPG = "JPG",
  PNG = "PNG",
  WEBP = "WEBP",
}

//==============================================================
// END Enums and Input Objects
//==============================================================
