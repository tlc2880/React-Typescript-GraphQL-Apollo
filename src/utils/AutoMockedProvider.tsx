//  Name: Tommy Cao
//  Date: 10/22/20
//  Title: Generating TypeScript Types from GraphQL Schema in Apollo

import React, { ReactNode } from "react";
import { ApolloProvider } from "react-apollo";
import { ApolloClient } from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import { SchemaLink } from "apollo-link-schema";
import {
  makeExecutableSchema,
  addMockFunctionsToSchema,
  IMocks
} from "graphql-tools";
import { printSchema, buildClientSchema } from "graphql/utilities";
import introspectionResult from "../../schema.json";

const AutoMockedProvider: React.FunctionComponent<{
  children: ReactNode;
  mockResolvers?: IMocks;
}> = ({ children, mockResolvers }) => {
  // 1) Convert JSON schema into Schema Definition Language
  const schemaSDL = printSchema(
    buildClientSchema({ __schema: introspectionResult.__schema as any })
  );

  // 2) Make schema "executable"
  const schema = makeExecutableSchema({
    typeDefs: schemaSDL,
    resolverValidationOptions: {
      requireResolversForResolveType: false
    }
  });

  // 3) Apply mock resolvers to executable schema
  addMockFunctionsToSchema({ schema, mocks: mockResolvers });

  // 4) Define ApolloClient (client variable used below)
  const client = new ApolloClient({
    link: new SchemaLink({ schema }),
    cache: new InMemoryCache()
  });

  return <ApolloProvider client={client}>{children}</ApolloProvider>;
};

export default AutoMockedProvider;
