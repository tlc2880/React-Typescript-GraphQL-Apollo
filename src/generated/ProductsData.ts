//  Name: Tommy Cao
//  Date: 10/22/20
//  Title: Generating TypeScript Types from GraphQL Schema in Apollo/

/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

import { ImageContentType } from "./globalTypes";

// ====================================================
// GraphQL query operation: ProductsData
// ====================================================

export interface ProductsData_products_edges_node_images_edges_node {
  __typename: "Image";
  /**
   * A unique identifier for the image.
   */
  id: string | null;
  /**
   * The location of the transformed image as a URL.
   * 
   * All transformation arguments are considered "best-effort". If they can be applied to an image, they will be.
   * Otherwise any transformations which an image type does not support will be ignored.
   */
  transformedSrc: ShopifyURL;
}

export interface ProductsData_products_edges_node_images_edges {
  __typename: "ImageEdge";
  /**
   * The item at the end of ImageEdge.
   */
  node: ProductsData_products_edges_node_images_edges_node;
}

export interface ProductsData_products_edges_node_images {
  __typename: "ImageConnection";
  /**
   * A list of edges.
   */
  edges: ProductsData_products_edges_node_images_edges[];
}

export interface ProductsData_products_edges_node {
  __typename: "Product";
  /**
   * Globally unique identifier.
   */
  id: string;
  /**
   * The product’s title.
   */
  title: string;
  /**
   * Stripped description of the product, single line with HTML tags removed.
   */
  description: string;
  /**
   * The date and time when the product was last modified.
   */
  updatedAt: ShopifyDateTime;
  /**
   * List of images associated with the product.
   */
  images: ProductsData_products_edges_node_images;
}

export interface ProductsData_products_edges {
  __typename: "ProductEdge";
  /**
   * The item at the end of ProductEdge.
   */
  node: ProductsData_products_edges_node;
}

export interface ProductsData_products {
  __typename: "ProductConnection";
  /**
   * A list of edges.
   */
  edges: ProductsData_products_edges[];
}

export interface ProductsData {
  /**
   * List of the shop’s products.
   */
  products: ProductsData_products;
}

export interface ProductsDataVariables {
  preferredContentType?: ImageContentType | null;
}
