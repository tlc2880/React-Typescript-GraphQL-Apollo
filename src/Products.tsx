//  Name: Tommy Cao
//  Date: 10/22/20
//  Title: Generating TypeScript Types from GraphQL Schema in Apollo

import React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import { ProductsData, ProductsDataVariables } from "./generated/ProductsData";
import { ImageContentType } from "./generated/globalTypes";

export const PRODUCTS_QUERY = gql`
  query ProductsData($preferredContentType: ImageContentType) {
    products(first: 10) {
      edges {
        node {
          id
          title
          images(first: 3) {
            edges {
              node {
                id
                transformedSrc(
                  maxWidth: 150
                  maxHeight: 100
                  preferredContentType: $preferredContentType
                )
              }
            }
          }
        }
      }
    }
  }
`;

class ProductsQuery extends Query<ProductsData, ProductsDataVariables> {}

export default function Products() {
  return (
    <ProductsQuery
      query={PRODUCTS_QUERY}
      variables={{ preferredContentType: ImageContentType.JPG }}
    >
      {({ data, loading, error }) => {
        if (error) {
          return <div>Error loading products...</div>;
        }
        if (loading || !data) {
          return <div>Loading products...</div>;
        }

        return (
          <div data-testid="result">
            <h1>Generating TypeScript Types from GraphQL Schema in Apollo</h1>
            {data.products.edges.map(({ node: product }) => (
              <div key={product.id}>
                <h2>{product.title}</h2>
                <p>ID {product.id}</p>

                <ul className="images">
                  {product.images.edges.map(
                    ({ node: image }, index: number) => (
                      <li className="image-item" key={image.id || index}>
                        <img src={image.transformedSrc} />
                      </li>
                    )
                  )}
                </ul>
              </div>
            ))}
          </div>
        );
      }}
    </ProductsQuery>
  );
}
