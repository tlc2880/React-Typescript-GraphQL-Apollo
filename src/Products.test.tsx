//  Name: Tommy Cao
//  Date: 10/22/20
//  Title: Generating TypeScript Types from GraphQL Schema in Apollo

import React from "react";
import { MockedProvider } from "react-apollo/test-utils";
import { render, cleanup } from "react-testing-library";
import { MockList } from "graphql-tools";
import faker from "faker";
import "jest-dom/extend-expect";
import { ImageContentType } from "./generated/globalTypes";
import AutoMockedProvider from "./utils/AutoMockedProvider";
import Products, { PRODUCTS_QUERY } from "./Products";

afterEach(cleanup);

const mocks = [
  {
    request: {
      query: PRODUCTS_QUERY,
      variables: {
        preferredContentType: ImageContentType.JPG
      }
    },
    result: {
      data: {
        products: {
          edges: [
            {
              node: {
                id: "123",
                title: "Nike Shoes",
                images: {
                  edges: [
                    {
                      node: {
                        id: "456",
                        transformedSrc: "https://www.images.com/shoe.jpg"
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
];

it("renders with MockedProvider", async () => {
  const { findByText, getByText } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Products />
    </MockedProvider>
  );

  expect(getByText("Loading products...")).toBeInTheDocument();

  const productTag = await findByText("Nike Shoes");
  expect(productTag).toBeInTheDocument();
});

it("renders with AutoMockedProvider", async () => {
  const mockResolvers = {
    Product: () => ({
      title: "Nike Shoes",
      images: () => ({
        edges: () => new MockList([0, 3])
      })
    }),
    Image: () => ({
      transformedSrc: (_: any, { preferredContentType }: any) =>
        `https://images.com/cat.${preferredContentType.toLowerCase()}`
    })
  };

  const { findByText, getByText } = render(
    <AutoMockedProvider mockResolvers={mockResolvers}>
      <Products />
    </AutoMockedProvider>
  );

  expect(getByText("Loading products...")).toBeInTheDocument();

  const productTag = await findByText("Nike Shoes");
  expect(productTag).toBeInTheDocument();
});

it("matches snapshot using seeds", async () => {
  faker.seed(123);

  const { findByTestId, asFragment } = render(
    <AutoMockedProvider
      mockResolvers={{
        URL: () => "https://www.shopify.com",
        ID: () => faker.random.uuid()
      }}
    >
      <Products />
    </AutoMockedProvider>
  );

  await findByTestId("result");
  expect(asFragment()).toMatchSnapshot();
});
