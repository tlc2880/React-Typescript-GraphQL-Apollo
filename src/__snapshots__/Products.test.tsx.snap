// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot using seeds 1`] = `
<DocumentFragment>
  <div
    data-testid="result"
  >
    <div>
      <h2>
        Hello World
      </h2>
      <p>
        ID bb463b8b-b76c-4f6a-9726-65ab5730b69b
      </p>
      <ul
        class="images"
      >
        <li
          class="image-item"
        >
          <img
            src="https://www.shopify.com"
          />
        </li>
        <li
          class="image-item"
        >
          <img
            src="https://www.shopify.com"
          />
        </li>
      </ul>
    </div>
    <div>
      <h2>
        Hello World
      </h2>
      <p>
        ID 7f3869c1-7dc9-4486-9045-6bade487a49d
      </p>
      <ul
        class="images"
      >
        <li
          class="image-item"
        >
          <img
            src="https://www.shopify.com"
          />
        </li>
        <li
          class="image-item"
        >
          <img
            src="https://www.shopify.com"
          />
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
